clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -1101,6 +1101,37 @@ def BitfieldInfoAttr : CIR_Attr<"BitfieldInfo", "bitfield_info"> {
   ];
 }
 
+//===----------------------------------------------------------------------===//
+// AnnotationAttr
+//===----------------------------------------------------------------------===//
+def AnnotationAttr : CIR_Attr<"Annotation", "annotation"> {
+  let summary = "Annotation attribute for global variables and functions";
+  let description = [{
+    Represent C/C++ attribute of annotate in CIR. example c code is:
+    ```
+     in test.c line 10
+     int *a __attribute__((annotate("testptr", "21", 12 ))); 
+    ```
+     this example code, the AnnotationAttr records annotation name "testptr", 
+     source file "test.c", line number 12, and arguments "21" and 12 for 
+     global variable a
+  }];
+
+  // Use ConstArrayAttr as type of name and translation unit
+  // because they will be used to generate global variable. Using ConstArrayAttr
+  // is a lot more suitable for that than StringAttr.
+  let parameters = (ins "ConstArrayAttr":$name,
+                        "ConstArrayAttr":$unit,
+                        "IntAttr":$lineno,
+                        "mlir::Attribute":$args);
+
+  let assemblyFormat = "`<` struct($name, $unit, $lineno, $args) `>`";                      
+
+  let extraClassDeclaration = [{
+    bool isNoArgs() const { return mlir::isa<mlir::cir::ZeroAttr>(getArgs()); };
```

Comment:
Since args should be `ArrayAttr`, this should be `bool isNoArgs() const { return getArgs(); };`. The fact that LLVM needs it to be finalized with a nullptr doesn't mean we need to do the same right now, we should just make lowering to put that extra nullptr there.